### PR TITLE
Fix edit issue

### PR DESCRIPTION
**Pull request by syohex**
This was introduced by #339.

## Related issue
https://github.com/emacsorphanage/helm-ag/issues/218#issuecomment-692311439

### CODE REVIEW

File: helm-ag.el [hunk 1]
```diff
@@ -624,7 +624,7 @@ Default behaviour shows finish and result in mode-line."
   (interactive)
   (goto-char (point-min))
   (let ((read-only-files 0)
-        (files-to-lines (make-hash-table))
+        (files-to-lines (make-hash-table :test #'equal))
```
Comment (syohex) [inline helm-ag.el:627]: The type of this hash key is list, so `equal` must be used as test function

File: helm-ag.el [hunk 2]
```diff
@@ -637,14 +637,16 @@ Default behaviour shows finish and result in mode-line."
         (if (not (file-writable-p file))
             (cl-incf read-only-files)
           (if lines-list
-              (push (list line body ovs) lines-list)
+              (progn
+                (push (list line body ovs) lines-list)
+                (puthash file lines-list files-to-lines))
```
Comment (syohex) [inline helm-ag.el:642]: In original code, updated value is not stored.

File: helm-ag.el [hunk 3]
```diff
@@ -637,14 +637,16 @@ Default behaviour shows finish and result in mode-line."
...
             (puthash file (list (list line body ovs)) files-to-lines)))))
     ;; Batch edits by file
     (maphash
      (lambda (curr-file lines-data)
        (with-temp-buffer
          (insert-file-contents curr-file)
-         (dolist (curr-line-data lines-data)
+         (dolist (curr-line-data (reverse lines-data))
```
Comment (syohex) [inline helm-ag.el:649]: I suppose file content should be updated from lower line to higher line.

**Comment by syohex** [comment]
@Namburgesas Could you review this patch as possible ?

**Comment by jcs090218** [comment]
Thanks! 😄

File: helm-ag.el (hunk 2 above)
Comment (Namburgesas) [reply helm-ag.el:642]: Ah lisp list...needed to update the map to point to the new head of the list.

File: helm-ag.el (hunk 1 above)
Comment (Namburgesas) [reply helm-ag.el:627]: Apologies, I understand now. Strings are objects too.

File: helm-ag.el (hunk 3 above)
Comment (Namburgesas) [reply helm-ag.el:649]: Yep, lisp beginner mistake.

**Comment by Namburgesas** [comment]
Thank you @syohex for fixing my bugs. It would appear that data is stored differently for `helm-do-ag-this-file` compared with `helm-dir-do-ag` and others. I had tested the latter in #339 .

**Comment by Namburgesas** [comment]
Looks like the latter is actually the spacemacs function `spacemacs/helm-dir-do-ag`(I use spacemacs), which ends up calling `helm-do-ag`. I'll look into this.

**Comment by Namburgesas** [comment]
We're good to go 👍

Root cause:

* `helm-do-ag-this-file` -> `helm-do-ag default-directory buffer-file-name`
Helm is configured with `search-this-file=t`.
For a given file, this results in the same string object representing the filename for each matching line, which surfaced the missing `puthash` bug.

* `spacemacs/helm-dir-do-ag` -> `helm-do-ag dir`
Helm is configured with `search-this-file=nil`.
For a given file, this results in a new string object representing the filename for each matching line. (Probably due to Helm's implementation of searching across multiple files.) This, coupled with `:test 'eql`, hid the issue when using `spacemacs/helm-dir-do-ag`.

**Pull request merged**